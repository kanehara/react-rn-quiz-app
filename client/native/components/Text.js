import {Text as _Text} from 'react-native'
import styled from 'styled-components'

const Text = styled(_Text)`
  font-size: 16px;
  margin: 4px 0 4px;
  color: ${({ theme }) => theme.fontColor};
`

export default Text
