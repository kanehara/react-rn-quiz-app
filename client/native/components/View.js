import styled from 'styled-components'

const View = styled.View`
  align-items: center;
  justify-content: center;
`

export const ScreenView = styled(View)`
  flex: 1;
  padding: 40px 20px;
  background-color: ${({theme}) => theme.backgroundColor};
`

export default View
