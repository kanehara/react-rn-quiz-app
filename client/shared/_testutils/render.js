import React from 'react'
import { ThemeProvider } from 'styled-components'
import theme from 'shared/theme'
import { Provider } from 'react-redux' 
import { store } from 'shared/redux'
import renderer from 'react-test-renderer'

export const render = jsx => renderer.create(
  <Provider store={store}>
    <ThemeProvider theme={theme}>
      {jsx}
    </ThemeProvider>
  </Provider>
)

