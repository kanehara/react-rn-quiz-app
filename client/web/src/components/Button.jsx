import styled from 'styled-components'

const Button = styled.button`
  padding: 10px;
  width: 100%;
  font-size: 18px;
  border-radius: ${({theme}) => theme.buttonBorderRadius};
  border: ${({theme}) => theme.buttonBorder};
  font-weight: bold;
  background-color: ${props => props.theme.buttonColor};

  &:hover {
    cursor: pointer;
  }
`

export const TrueButton = styled(Button)`
  background-color: ${({theme}) => theme.trueButtonColor};
`

export const FalseButton = styled(Button)`
  background-color: ${({theme}) => theme.falseButtonColor};
`

export default Button
