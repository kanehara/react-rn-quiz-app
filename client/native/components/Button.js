import styled from 'styled-components'
import {View} from 'react-native'
import Text from 'native/components/Text'
import React from 'react'

const TouchableHighlight = styled.TouchableHighlight`
  background-color: ${({ theme }) => theme.buttonColor};
  border-radius: ${({theme}) => theme.buttonBorderRadius};
  border: ${({theme}) => theme.buttonBorder};
  
  width: 100%;
  align-items: center;
  justify-content: center;
  padding: 5px 10px;
  margin: 10px;
`

const ButtonText = styled(Text)`
  font-weight: 700;
`

const Button = props => (
  <TouchableHighlight 
    style={props.style} 
    onPress={props.onPress}
    underlayColor='rgba(0, 0, 0, 0.15)'
  >
    <View>
      <ButtonText>{props.title}</ButtonText>
    </View>
  </TouchableHighlight>
)

export const FalseButton = styled(Button)`
  background-color: ${({theme}) => theme.falseButtonColor};
`

export const TrueButton = styled(Button)`
  background-color: ${({theme}) => theme.trueButtonColor};
`

export default Button
