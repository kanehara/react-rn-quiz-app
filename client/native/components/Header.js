import Text from 'native/components/Text'
import styled from 'styled-components'

const H = styled(Text)`
  text-align: center;
  font-weight: bold;
`

export const H1 = styled(H)`
  font-size: 28px;
  margin-bottom: 16px;
`

export const H2 = styled(H)`
  font-size: 22px;
  margin-bottom: 12px;
`

export const H3 = styled(H)`
  font-size: 18px;
  margin-bottom: 10px;
`

export const H4 = styled(H)`
  font-size: 16px;
  margin-bottom: 8px;
`

export const H5 = styled(H)`
  font-size: 12px;
  margin-bottom: 6px;
`
