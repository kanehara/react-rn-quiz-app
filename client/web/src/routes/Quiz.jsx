import React from 'react'
import { Ray, Flex } from "rayout"
import { connect } from 'react-redux'
import { actions, getters } from 'shared/redux/quiz'
import {Redirect} from 'react-router-dom'
import {TrueButton as _TrueButton, FalseButton} from '../components/Button'
import styled from 'styled-components'

const TrueButton = styled(_TrueButton)`
  margin-bottom: 35px;
`

const Card = styled.div`
  padding: 25px;
  background-color: white;
  border-radius: 15px;
  box-shadow: 0px 5px 18px 5px #e5e5e5;
  width: 95%;

  @media (min-width: 600px) {
    width: 80%;
  }
  
  @media (min-width: 1000px) {
    width: 50%;
  }
`

class Quiz extends React.Component {
  get question () {
    return this.props.questions[this.props.questionIndex]
  }

  render () {
    return this.props.finished
      ? <Redirect to='/results' />
      : this.question
        ? (
          <Card display='flex'>
            <h3>{this.question.category}</h3>
            <Ray minHeight='175px' marginBottom='20px'>
              <h1>{this.question.question}</h1>
              <h5>{this.props.questionIndex + 1} of {this.props.questions.length}</h5>
            </Ray>
            <Flex flexDirection='column'>
              <TrueButton onClick={() => this.props.setAnswer('True')}>True</TrueButton>
              <FalseButton onClick={() => this.props.setAnswer('False')}>False</FalseButton>
            </Flex>
          </Card>
        )
        : (
          <Redirect to='/' />
        )
  }
}

export default connect(
  state => ({
    finished: getters.finished(state),
    questions: state.questions,
    questionIndex: state.questionIndex
  }),
  dispatch => ({
    setAnswer: answer => dispatch(actions.setAnswer(answer))
  })
)(Quiz)
