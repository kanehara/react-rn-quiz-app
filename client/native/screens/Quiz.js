import React from 'react'
import {H1, H3, H5} from 'native/components/Header'
import View, {ScreenView} from 'native/components/View'
import {TrueButton, FalseButton} from 'native/components/Button'
import { connect } from 'react-redux'
import { actions, getters } from 'shared/redux/quiz'
import styled from 'styled-components'

const Card = styled(View)`
  padding: 25px 15px;
  background-color: white;
  border-radius: 15px;
  border: 1px solid #e5e5e5;
  /* box-shadow: 0px 5px 18px 5px #e5e5e5; */
  align-items: center;
  justify-content: center;
  width: 95%;
`

class QuizScreen extends React.Component {
  get question () {
    return this.props.questions[this.props.questionIndex] || {}
  }

  componentDidUpdate() {
    if (this.props.finished) {
      this.props.navigation.navigate('results')
    } else if (!this.question) {
      this.props.navigation.navigate('home')
    }
  }

  handleAnswer = answer => {
    this.props.setAnswer(answer)
    this.props.navigation.navigate({
      routeName: 'quiz',
      params: { index: this.props.questionIndex + 1 }
    })
  }

  render() {
    return (
      <ScreenView>
        <Card>
          <H3>{this.question.category}</H3>
          <H1>{this.question.question}</H1>
          <H5>{this.props.questionIndex + 1} of {this.props.questions.length}</H5>
          <TrueButton onPress={() => this.handleAnswer('True')} title='True' />
          <FalseButton onPress={() => this.handleAnswer('False')} title='False' />
        </Card>
      </ScreenView>
    );
  }
}

export default connect(
  state => ({
    finished: getters.finished(state),
    questions: state.questions,
    questionIndex: state.questionIndex
  }),
  dispatch => ({
    setAnswer: answer => dispatch(actions.setAnswer(answer))
  })
)(QuizScreen)
