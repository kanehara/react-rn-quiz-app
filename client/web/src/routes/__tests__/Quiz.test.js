import {Quiz, TrueButton} from '../Quiz'
import React from 'react'
import { render } from 'shared/_testutils'
import {MemoryRouter} from 'react-router-dom'
import {FalseButton} from 'web/src/components/Button'

describe('Quiz', () => {
  it('renders a redirect to / if no question', () => {
    expect(render(
      <MemoryRouter>
        <Quiz questions={[{}]} questionIndex={3} />
      </MemoryRouter>
    )).toMatchSnapshot()
  })
  
  it('renders a redirect to /results if finished', () => {
    expect(render(
      <MemoryRouter>
        <Quiz finished />
      </MemoryRouter>
    )).toMatchSnapshot()
  })

  it('renders question if present', () => {
    const setAnswerMock = jest.fn()
    const instance = render(
      <Quiz 
        questionIndex={0}
        setAnswer={setAnswerMock}
        questions={[{
          category: 'Rick & Morty',
          question: 'Who is Rick?'
        }]} />
    )
    expect(instance).toMatchSnapshot()

    instance.root.findByType(TrueButton).props.onClick()
    expect(setAnswerMock).toHaveBeenLastCalledWith('True')
    instance.root.findByType(FalseButton).props.onClick()
    expect(setAnswerMock).toHaveBeenLastCalledWith('False')
  })
})
