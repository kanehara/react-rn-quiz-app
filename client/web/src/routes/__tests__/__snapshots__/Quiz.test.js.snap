// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz renders a redirect to / if no question 1`] = `
<div
  data-testid="homeRedirect"
/>
`;

exports[`Quiz renders a redirect to /results if finished 1`] = `
<div
  data-testid="resultsRedirect"
/>
`;

exports[`Quiz renders question if present 1`] = `
<div
  className="sc-ifAKCX bTtQcI"
  display="flex"
>
  <h3>
    Rick & Morty
  </h3>
  <div
    className="sc-bdVaJa lgpJss"
  >
    <h1>
      Who is Rick?
    </h1>
    <h5>
      1
       of 
      1
    </h5>
  </div>
  <div
    className="sc-bwzfXH giqBiD"
  >
    <button
      className="sc-bdVaJa sc-bwzfXH sc-bxivhb jJnRuO"
      onClick={[Function]}
    >
      True
    </button>
    <button
      className="sc-bdVaJa sc-htpNat kAvMjd"
      onClick={[Function]}
    >
      False
    </button>
  </div>
</div>
`;
